src/Common/MemorySpillScheduler.cpp

```diff
@@ -0,0 +1,69 @@
+#include <mutex>
+#include <Common/MemoryTracker.h>
+#include <Common/MemoryTrackerUtils.h>
+#include <Common/MemorySpillScheduler.h>
+#include <Processors/IProcessor.h>
+
+#include <Poco/Logger.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+void MemorySpillScheduler::checkAndSpill(IProcessor * processor)
+{
+
+    if (!enable || !processor->spillable() || !getHardLimit())
+    {
+        return;
+    }
+    auto stats = processor->getMemoryStats();
+    std::lock_guard lock(mutex);
+    processor_stats[processor] = stats;
+    updateTopProcessor();
+    if (getCurrentQueryMemoryUsage() + max_reserved_memory_bytes < getHardLimit())
+        return;
+    LOG_DEBUG(getLogger("MemorySpillScheduler"),
+        "need to spill. current memory usage: {}, hard limit: {}, max_reserved_memory_bytes: {}",
+        getCurrentQueryMemoryUsage(), getHardLimit(), max_reserved_memory_bytes);
+    if (processor == top_processor)
+        processor->spillOnSize(stats.spillable_memory_bytes);
```

Comment:
Should we update the stats under `lock` after we are done with spilling? Another processor might be waiting on the lock and redo spilling with old stats, leading to the wrong top processor. It would be great to cover this case in tests as well if you can come up with a simple test.

Reply:
`spillOnSize` may not actully to the spill, like `FillingRightJoinSideTransform` does, it just sets a flag. So 
1. after `spillOnSize` is called, the memory may still the same, like `FillingRightJoinSideTransform` does. In most cases, we expect the processor to release memory at least after `work`.
2. If the memory statistics of processor selected to spill is not updated (it will be until next round of scheduling the processor to work), there will not be any other processor be selected to spill.

Reply:
I think I make a mistake before， `spillOnSize` should be put under the `lock`, it may block other processors to long.